base/floatfuncs.jl

```diff
@@ -16,6 +16,9 @@ signbit(x::Float64) = signbit(bitcast(Int64, x))
 signbit(x::Float32) = signbit(bitcast(Int32, x))
 signbit(x::Float16) = signbit(bitcast(Int16, x))
 
+ispositive(x::IEEEFloat) = x>0
+isnegative(x::IEEEFloat) = x<0
```

Comment:
I would suggest we make these the default implementations. They are much more intuitive. If the point of this PR is to make `BigFloat` faster, focus on that.

Reply:
The point is that `ispositive` and `isnegative` can work without comparison operators. For a custom type without hardware support, it's usually easier and faster to implement `signbit` and `iszero`.

`BigInt` and `isnegative(::Unsigned)` should also be faster.

Reply:
These functions only make sense for `Real` anyway (and one could argue that they maybe be restricted as such? I'd suggest this except for the `Dual` type below.). Is it a realistic case that you implement a `Real` type where you don't care about comparisons but *do* care about signs? Both seems pretty fundamental to real numbers.

Is `signbit` really a fundamental function? More-so than comparison? Note that the most general definition is `signbit(x::Real) = x < 0`, so it uses comparison unless given an explicitly better option. In the meantime,
```julia-repl
julia> using DualNumbers, BenchmarkTools

julia> ltzero(x) = x<0
ltzero (generic function with 1 method)

julia> isnegative(x) = !isnan(x) && signbit(x) && !iszero(x)
isnegative (generic function with 1 method)

julia> d = Dual.(randn(1000));

julia> @btime count(ltzero, $d)
  179.714 ns (0 allocations: 0 bytes)
491

julia> @btime count(isnegative, $d)
ERROR: MethodError: no method matching signbit(::Dual128)
```
Now, maybe `Dual` should support `signbit` (for wrapped `Real` types, although generally `Dual <: Number` so it doesn't make sense). But here's a living example from the ecosystem that would not work with a `signbit` implementation.

`Unsigned` is an abstract type, but for any `BitUnsigned` you'll get
```julia-repl
julia> code_llvm(x -> x < 0, (UInt,))
;  @ REPL[35]:1 within `#27`
define i8 @"julia_#27_348"(i64 zeroext %0) #0 {
top:
  ret i8 0
```
so this is no help. If you want to make `BigInt` faster, go ahead and specialize them too.

Reply:
> Is it a realistic case that you implement a `Real` type where you don't care about comparisons but _do_ care about signs?

Some types are not easy/able to compare but we can easily check that they are positive or negative. For example, irrationals (that's my motivation) and dates (from https://github.com/JuliaLang/julia/issues/35067).

On the second case, it's suggested to use `x>zero(x)` and I think this should be the general fallback. For `Real` it should just be `x>0`.

> Is signbit really a fundamental function?

It should be on the same level as `iszero`. Its name might be bad, but the fallback suggests it's basically an `almost_isnegative` and its documentation says
`Return true if the value of the sign of x is negative, otherwise false.`

Reply:
`signbit` is identical to `<(0)` except for its treatment of "negative zero" and NaN. As such, I consider it to be a function for bit-twiddling or explicitly checking the sign *bit* (where one wants to group +0 and +NaN with the positives and -0 and -NaN with the negatives, except that one usually doesn't actually care about the signs of NaNs). In other words, I consider `signbit` to be a statement about implemented representation rather than value.

I consider `sign` or `<(0)` to be "mathematical" functions that concern values rather than representation. Although `sign` still distinguishes +0 and -0, both are usually `==(0)`. I acknowledge that a version of `isnegative` for `BigInt` and maybe some other types are currently more efficient with `signbit`, but that's because of internal representation (which is why I advocate a specialization for such cases). Note that if we did (I doubt we would on any <5yr timescale) replace the `BigInt`/`BigFloat` libraries with native Julia implementations that handle mixed-type comparisons, this performance difference might reverse.

So personally, I consider `signbit` (a representation-testing function that happens to align with a value-testing function for some types) to be subordinate to meaning-based tests like `iszero` or `sign` or `<`. When writing operational code, I contend that one is more interested in meaning than representation. This is why I prefer `x -> x<0` as the generic implementation rather than the `signbit` implementation.

Note: although I often write `<(0)`, one would usually implement this as `x -> x<0` because `<(0)` is a `Base.Fix2` type and that often blocks propagation of the constant.